### PR TITLE
Microphysics emulation model utilities

### PR DESCRIPTION
**Pull request by frodre**
This PR adds microphysics keras model creation to `fv3fit.  It can reproduce the tendency model types investigated by @jacnugent over the summer, but includes all the convenience tooling for directly outputting incremented fields from those predictions.  

Added public API:
- `MicrophysicsConfig` in `fv3fit.emulation.models`: model configuration/builder
- Many general use keras layers for model building in `fv3fit.emulation.layers`: 
  - `IncrementStateLayer`: Convenience layer for incrementing an input field tensor with a tendency
  - `IncrementedFieldOutput`: A layer that forces learned outputs to be tendency-based by incrementing an input field with denormalized model output.
    - Allows for retrieval of tendency outputs as well
  - `FieldInput`: Input layer that normalizes a field and optionally subselects along the feature dim
  - `FieldOutput`: Direct prediction of an output field using dense output w/ linear activation and denormalization.
  - `CombineInputs`: Concatenate input tensors along a dimension with the option to add a new dimension prior to combining
  - `RNNBlock`: A simple RNN architecture layer with optional connected dense layers following
  - `MLPBlock`: A simple dense network block with 0-many hidden layers

Significant internal changes:
- `TransformConfig` in `fv3fit.emulation.data` now includes a `DerivedMapping` transform before conversion to arrays/tensors

- [x ] Tests added

### CODE REVIEW

File: external/fv3fit/fv3fit/emulation/layers/normalization.py [hunk 1]
```diff
@@ -165,3 +169,23 @@ class MeanFeatureStdNormLayer(MaxFeatureStdNormLayer, FeatureMeanStd):
     """
 
     pass
+
+
+def get_norm_class(key):
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/layers/normalization.py:174]: Based on your comment, the idea is instantiated layers returned from these classes.  I think I might just remove them from the public API and move the function to `models.py` since that's where I'm using it.

File: external/vcm/vcm/derived_mapping.py [hunk 2]
```diff
@@ -242,3 +247,36 @@ def pQ2(self):
...
+    end = self._mapper["air_temperature_output"]
+    begin = self._mapper["air_temperature_input"]
+    return (end - begin) / self._microphys_timestep_sec
+
+
+@DerivedMapping.register(
+    "tendency_of_specific_humidity_due_to_microphysics",
+    required_inputs=["specific_humidity_input", "specific_humidity_output"],
+)
+def tendency_of_specific_humidity_due_to_microphysics(self):
```
Comment (AnnaKwa) [inline external/vcm/vcm/derived_mapping.py:266]: Just an FYI, I'm not sure how the emulation branch plugs these into the prognostic run, but now the master version of the prognostic run code relies on hard coded mappings of tendency names to the state variable they update (https://github.com/ai2cm/fv3net/blob/a3fc28bf21eaa5ed9b2ca2a2a8cdeb4c4748e8f2/workflows/prognostic_c48_run/runtime/names.py#L22), so if the emulation models want to plug in that way in the future you just need to add the appropriate key/value pair for the prediction variables.

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks, there is a lot of useful stuff in here! It feels like the building blocks of the emulation models are similar enough to the other non-emulation keras code that it wouldn't be too hard in the future to use this outside of emulation. 

My main high level comment is that reading through the layers module, it felt like it was more general than just emulation and could live in the keras directory. I think the closer together the emulation/non-emulation keras models live, the more likely we can make use of new stuff from either side and they'll converge as we develop more as oppose to drifting farther apart. Since the emulation/non-emulation models still use separate modules though that could make the keras model directory harder to navigate. I'll leave it up to you as to where to put the layers code.

File: external/fv3fit/fv3fit/emulation/layers/util.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    Attributes:
+        dt_sec: timestep delta in seconds
+    """
+
+    def __init__(self, dt_sec: int, *args, dtype=tf.float32, **kwargs):
+
+        self.dt_sec = tf.constant(dt_sec, dtype=dtype)
+        super().__init__(*args, **kwargs)
+
+    def call(self, tensors: Sequence[tf.Tensor]) -> tf.Tensor:
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/layers/util.py:18]: Ah no there isn't a good reason.  I think I was basing this off of arbitrary number of input layers like Add, which clearly never applies here. Ty!

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        tensor = self.unscaled(tensor)
+        tensor = self.denorm(tensor)
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+
+class ResidualOutput(tf.keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:93]: Yep, went back and forth on this myself.  Updated to `IncrementedFieldOutput`

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        field_input, network_output = tensors
+        tendency = self.tendency(network_output)
+        tensor = self.increment([field_input, tendency])
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+    def get_tendency_output(self, network_output):
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:140]: Yes, it could be removed.  I think the current style is a bit more expressive on the dev side since it explicitly calls out what is being grabbed and what the expected input is.  Though, I should probably make the layer attributes "_"-prefixed to express that they are private.

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        super().__init__(*args, **kwargs)
+
+        self.combine_axis = combine_axis
+        self.expand = expand
+
+    def call(self, inputs):
+
+        if self.expand:
+            inputs = [
+                tf.expand_dims(tensor, axis=self.combine_axis) for tensor in inputs
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:157]: Good point of potential confusion.  I switched the layer init to specify an explicit expansion axis and added docstrings about the args specifying the order of operations.

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        """
+        Args:
+            width: width of MLP layer, only used if dense_depth > 0
+            depth: number of MLPlayers, set to 0 to disable
+            combine_inputs: whether to stack the inputs into a single block
+            activation: activation function to use for RNN and MLP
+        """
+        super().__init__(*args, **kwargs)
+
+        self.combine = _combine_initializer(combine_inputs, -1, False, self.name)
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:240]: I was initially thinking that handling the inputs was an implementation detail of the block layers, but I agree that it makes things repetitive without much benefit.  I'm moving this to the model building and making the layers just accept a single tensor.

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        return outputs
+
+
+class LinearBlock(MLPBlock):
+    """
+    No special prediction architecture, just combine input features
+    for connection to outputs.
+    """
+
+    def __init__(self, *args, combine_inputs=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:262]: Yes, this was more of a layer name for the summary with combining layer but now I have removed in favor of the model builder combining.

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:1]: I like this and thanks for pointing me to the similar features in the keras space.   I think any refactors/unification should be in it's own PR, so I'll talk with @mcgibbon to see what he thinks on this too.

File: external/fv3fit/fv3fit/emulation/microphysics/layers.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+class MLPBlock(tf.keras.layers.Layer):
+    """MLP layer for basic NN predictions"""
+
+    def __init__(
+        self,
+        *args,
+        width: int = 256,
+        depth: int = 2,
+        combine_inputs: bool = True,
+        activation: str = "relu",
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/layers.py:221]: There's the subtle difference of built tensor trace vs. a layer used to build a tensor trace, but we should be able to facilitate both if we deside whether layers work as a nice base unit for model building.  Since the args here are nearly a subset of the config, I think I'll leave as is for now and handle in the other PR if we decide to go that route.

File: external/vcm/vcm/derived_mapping.py (hunk 2 above)
Comment (frodre) [reply external/vcm/vcm/derived_mapping.py:266]: Thanks for noting this.  Currently, the emulator plugs in through a different mechanism that doesn't interact with the wrapper state, but I'll keep this in mind for our other all-physics emulation work.

**Comment by frodre** [comment]
Okay, ready for re-review.  I ended up moving a lot of the layers into the public emulation API and agree that it would be good to try and unify some of these with fv3fit in-general, but I think that's best left to another PR.

File: external/fv3fit/fv3fit/emulation/layers/fields.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+    def get_tendency_output(self, network_output):
+        return self.tendency(network_output)
+
+
+class CombineInputs(tf.keras.layers.Layer):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/layers/fields.py:167]: Seeing as how this is instantiated by a check on the architecture type, I think this class would be more at home in `layers.architecture` than `layers.fields`.

File: external/fv3fit/fv3fit/emulation/microphysics/models.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    """
+
+    name: str
+    kwargs: Mapping[str, Any] = dataclasses.field(default_factory=dict)
+
+    @property
+    def build(self):
+        return get_architecture_cls(self.name, kwargs=self.kwargs)
+
+
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/microphysics/models.py:52]: Up to this point in the module the classes aren't that specific to the microphysics case- instead of structuring the emulation directory like `emulation/microphysics/models`, if you foresee all-physics or non-microphysics emulation models being composed out of these blocks it might make sense to do `emulation/models` where that subdir has an architecture module (containing the stuff above this comment) as well as a microphysics module (containing the microphysics specific config class here). I'll leave that decision to you depending on your future plans for emulation models.

File: external/fv3fit/fv3fit/emulation/microphysics/models.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+            for name, sample in zip(self.input_variables, sample_in)
+        ]
+        residual_map = {
+            resid_name: inputs[self.input_variables.index(input_name)]
+            for resid_name, input_name in self.residual_out_variables.items()
+        }
+        processed = self._get_processed_inputs(sample_in, inputs)
+        combine_layer = get_combine_from_arch_key(self.architecture.name)
+        combined = combine_layer(processed)
+        arch_layer = self.architecture.build
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/microphysics/models.py:197]:
```suggestion
        arch_layer = self.architecture.build()
```

**Review by AnnaKwa — APPROVED** [review]
Thanks for the revisions, I had a couple minor comments and a suggestion for module organization (I'll leave that as your call) but don't need to re-review.

File: external/fv3fit/fv3fit/emulation/layers/architecture.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    which recurses backwards (top-to-bottom for the physics-param data)
+    by default over the z dimension.
+    """
+
+    def __init__(
+        self,
+        *args,
+        channels: int = 256,
+        dense_width: int = 256,
+        dense_depth: int = 1,
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/layers/architecture.py:18]: I'll opt to adjust the language in the docstring.  I want this to be the default for easy config of runs.

File: external/fv3fit/fv3fit/emulation/microphysics/models.py (hunk 13 above)
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/models.py:197]: This used to be a `@property` field of the class as `instance`.  I forgot to switch it over.  I'll make this adjustment in a single commit.

File: external/fv3fit/fv3fit/emulation/microphysics/models.py (hunk 12 above)
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/microphysics/models.py:52]: I think this makes a lot of sense for discoverability!  I moved the microphysics folder into `emulation/models` with the model in `microphysics.py` and the architecture related classes into `_core.py` since those should just be used for model building.
